File: crawler/company_industry.py
"""公司 → 行业 确定性分类器（与 lib/company-industry.js 同口径）。

爬虫端「行业-公司-岗位」跨行业门用：发现/刷新时按用户目标行业收窄，丢弃跨行业岗。
- 公司表（COMPANY_OVERRIDES）是**唯一数据源**，活在 JS（lib/company-industry.js），
  本模块读其生成产物 lib/data/company-industry-overrides.json（避免两份漂移；
  改 JS 后跑 `node scripts/gen-company-overrides-json.js` 重生成，sync 测试守卫）。
- 关键词规则 / 用户行业别名 / 门逻辑 按 china_keyword_expansion.py 惯例手镜像 JS（小、低频）。
"""
import json
import os
import re
from typing import List, Optional, Set

_OVERRIDES_PATH = os.path.join(
    os.path.dirname(__file__), "..", "lib", "data", "company-industry-overrides.json"
)

# 须与 lib/industries.ts 的 INDUSTRIES 真行业部分同口径（不含 央国企/其他）。
INDUSTRY_CATEGORIES = [
    "互联网/科技", "金融", "消费/零售", "制造/工业", "汽车/出行",
    "医疗/医药", "能源/化工", "地产/建筑", "物流/供应链", "传媒/文娱", "教育",
]


def _load_overrides():
    try:
        with open(_OVERRIDES_PATH, "r", encoding="utf-8") as f:
            return [(str(name), str(cat)) for name, cat in json.load(f)]
    except (OSError, ValueError):
        return []


COMPANY_OVERRIDES = _load_overrides()

# 行业关键词规则（公司名含该词 → 行业）。顺序敏感，与 JS INDUSTRY_KEYWORD_RULES 同口径。
_INDUSTRY_KEYWORD_RULES = [
    # ⚠️ 与 JS 逐字同口径（tests/company-industry-cross-lang.test.js 跨语言对拍守卫）。
    # 英文词一律带前词边界：不加会让 "express" 吃掉 American Express、"tech" 吃掉 BioNTech/Genentech、
    # "capital" 吃掉 CapitaLand。中文没有词边界概念，靠 COMPANY_OVERRIDES 的最长匹配兜。
    ("金融", re.compile(r"银行|证券|保险|基金|信托|期货|资管|财险|寿险|金融|支付|消费金融|小额贷|\bbank|\bsecurities\b|\binsurance\b|\bcapital\b", re.I)),
    ("医疗/医药", re.compile(r"医药|制药|药业|药品|生物医药|生物科技|医疗|医院|健康|基因|诊断|器械|\bpharma|\bbiotech\b|\bmedical\b|\bhealth", re.I)),
    ("汽车/出行", re.compile(r"汽车|整车|车业|新能源车|乘用车|商用车|车联网|出行|\bmotors\b|\bautomotive\b", re.I)),
    ("能源/化工", re.compile(r"能源|电力|电网|石油|石化|化工|化学|新能源|光伏|风电|储能|电池|燃气|煤业|核电|\benergy\b|\bpower\b|\bchemical|\bpetro", re.I)),
    ("物流/供应链", re.compile(r"物流|快递|供应链|仓储|货运|运输|冷链|\blogistics\b|\bexpress\b|\bsupply\s*chain\b", re.I)),
    ("地产/建筑", re.compile(r"地产|置业|房产|建筑|建设|建工|工程局|装饰|幕墙|\breal\s*estate\b|\bconstruction\b|\bpropert", re.I)),
    ("教育", re.compile(r"教育|学校|培训|学院|课程|留学|\beducation\b|\bacademy\b", re.I)),
    ("传媒/文娱", re.compile(r"传媒|影视|文化|娱乐|院线|音乐|动漫|文娱|出版|\bmedia\b|\bentertainment\b", re.I)),
    ("消费/零售", re.compile(r"食品|饮料|乳业|乳品|零售|商超|百货|便利店|美妆|化妆品|日化|服饰|服装|鞋业|家居|家电|餐饮|连锁|消费|快消|\bretail\b|\bconsumer\b|\bfoods?\b|\bbeverage\b", re.I)),
    ("制造/工业", re.compile(r"制造|机械|重工|工业|装备|设备|电子|半导体|芯片|集成电路|材料|钢铁|有色|精密|模具|纺织|轻工|\bmanufactur|\bindustrial\b|\belectronics\b|\bsemiconductor\b", re.I)),
    # ⚠️「科技」「智能」是中文公司名的通用后缀，各行各业都在用，拿它们判互联网 live 实测
    # 误判 162 家 / 10,757 个在招岗。已移除，判不出（None）走放行，不误杀。
    ("互联网/科技", re.compile(r"互联网|网络|信息技术|软件|数码|大数据|云计算|游戏|网游|人工智能|物联网|\btech\b|\bsoftware\b|\bdigital\b|\binternet\b|\bai\b|\bcloud\b", re.I)),
]
# 用户自填行业（自由文本）→ 规范类目。与 JS USER_INDUSTRY_ALIASES 同口径。
_USER_INDUSTRY_ALIASES = [
    ("互联网/科技", re.compile(r"互联网|科技|信息技术|软件|计算机|it|tech|游戏|人工智能|\bai\b|大数据|云", re.I)),
    ("金融", re.compile(r"金融|银行|证券|保险|基金|投资|fintech|finance", re.I)),
    ("消费/零售", re.compile(r"消费|零售|快消|fmcg|电商|食品|饮料|美妆|服装|retail|consumer", re.I)),
    ("制造/工业", re.compile(r"制造|工业|机械|电子|半导体|芯片|材料|硬件|manufactur|industrial", re.I)),
    ("汽车/出行", re.compile(r"汽车|车|出行|新能源车|automotive", re.I)),
    ("医疗/医药", re.compile(r"医疗|医药|生物|制药|健康|器械|pharma|bio|medical|health", re.I)),
    ("能源/化工", re.compile(r"能源|电力|化工|化学|新能源|光伏|电池|energy|chemical", re.I)),
    ("地产/建筑", re.compile(r"地产|房地产|建筑|建设|工程|real\s*estate|construction", re.I)),
    ("物流/供应链", re.compile(r"物流|供应链|快递|运输|logistics|supply", re.I)),
    ("传媒/文娱", re.compile(r"传媒|文娱|影视|文化|娱乐|内容|media|entertainment", re.I)),
    ("教育", re.compile(r"教育|培训|edu", re.I)),
]


def _normalize_company(value) -> str:
    return re.sub(r"\s+", " ", str(value or "").lower()).strip()


def _override_hits(text: str, key: str) -> bool:
    """override 单条是否命中。纯 ASCII 的 key 必须成词，中文 key 用 substring。
    与 JS overrideHits 同口径：不加词边界的话 "abb"(ABB) 会吃掉 "雅培 Abbott" 的单词内部。"""
    if re.fullmatch(r"[a-z0-9][a-z0-9 .&-]*", key):
        return re.search(r"(?<![a-z0-9])%s(?![a-z0-9])" % re.escape(key), text, re.I) is not None
    return key in text


def classify_company_industry(company) -> Optional[str]:
    """公司 → 行业类目（或 None=判不出）。overrides 优先于关键词规则。

    ⚠️ overrides 取**最长**命中，不是第一个命中：短品牌名会吃掉带它的长公司名
    （"京东"吃"京东方"、"网易"吃"网易云音乐"、"腾讯"吃"腾讯音乐"）。与 JS classifyCompanyIndustry 同口径。
    """
    text = _normalize_company(company)
    if not text:
        return None
    best_key, best_cat = None, None
    for name, cat in COMPANY_OVERRIDES:
        key = _normalize_company(name)
        if not _override_hits(text, key):
            continue
        if best_key is None or len(key) > len(best_key):
            best_key, best_cat = key, cat
    if best_cat:
        return best_cat
    for cat, rule in _INDUSTRY_KEYWORD_RULES:
        if rule.search(text):
            return cat
    return None


def canonicalize_user_industry(value) -> Optional[str]:
    text = _normalize_company(value)
    if not text:
        return None
    if value in INDUSTRY_CATEGORIES:
        return value
    for cat, rule in _USER_INDUSTRY_ALIASES:
        if rule.search(text):
            return cat
    return None


def user_target_industry_categories(industries) -> Set[str]:
    out: Set[str] = set()
    for raw in industries or []:
        cat = canonicalize_user_industry(raw)
        if cat:
            out.add(cat)
    return out


def job_industry_allowed(company, industries) -> bool:
    """跨行业门：放行当 用户没填可识别行业 / 岗位行业判不出 / 行业 ∈ 用户目标集合；
    拦截仅当 用户有明确目标行业 且 岗位行业已知 且 不在目标集合内。与 JS jobIndustryAllowed 同口径。"""
    targets = user_target_industry_categories(industries)
    if not targets:
        return True
    job_cat = classify_company_industry(company)
    if not job_cat:
        return True
    return job_cat in targets

File: lib/company-industry.js
// 公司 → 行业 确定性分类器（求职雷达「行业-公司-岗位」三层认知的地基，见记忆
// job-radar-industry-company-position-model）。岗位的行业从其**公司**派生：打分时
// 现算（job.company 本就带），零跨库、零迁移、零 API、可单测；算不出 → null → 放行（不误杀）。
//
// 设计：
//   1. 输出 = 下方 INDUSTRY_CATEGORIES 之一（须与 lib/industries.ts 的 INDUSTRIES 对齐）或 null。
//   2. 先查「大厂名映射」（substring，治名字不带行业词的品牌，如 农夫山泉/字节）→ 再走「关键词规则」。
//   3. 用户自填行业（candidate_profiles.industries 自由文本）经 canonicalizeUserIndustry 归一到同一类目空间再比对。
//   4. 保守：岗位行业未知 或 用户没填目标行业 → 不设门（放行），不拿缺数据误杀（沿用项目「信息不足放行」原则）。

// 须与 lib/industries.ts 的 INDUSTRIES 同口径（那是 .ts、node --test 不可 require，故此处镜像；改一处同步另一处）。
const INDUSTRY_CATEGORIES = [
  "互联网/科技",
  "金融",
  "消费/零售",
  "制造/工业",
  "汽车/出行",
  "医疗/医药",
  "能源/化工",
  "地产/建筑",
  "物流/供应链",
  "传媒/文娱",
  "教育",
];

function normalizeCompany(value) {
  return String(value || "").toLowerCase().replace(/\s+/g, " ").trim();
}

// 大厂/知名品牌名 → 行业（substring 命中，优先于关键词规则）。
// 治「名字不含行业词」的品牌（农夫山泉/字节/蔚来…）+ 纠正关键词会误判的（京东含『东』无意义、比亚迪是车不是泛制造）。
// 按 DB 实际产出（偏 车厂/制造/央企 + 外企）挑高频高价值公司；覆盖不全无妨——未命中走关键词、再不行 null 放行。
const COMPANY_OVERRIDES = [
  // 互联网/科技
  ["字节跳动", "互联网/科技"], ["抖音", "互联网/科技"], ["腾讯", "互联网/科技"], ["阿里巴巴", "互联网/科技"],
  ["阿里云", "互联网/科技"], ["蚂蚁集团", "金融"], ["百度", "互联网/科技"], ["美团", "互联网/科技"],
  ["快手", "互联网/科技"], ["拼多多", "互联网/科技"], ["京东", "互联网/科技"], ["网易", "互联网/科技"],
  ["小红书", "互联网/科技"], ["哔哩哔哩", "互联网/科技"], ["bilibili", "互联网/科技"], ["微博", "互联网/科技"],
  ["携程", "互联网/科技"], ["华为", "互联网/科技"], ["小米", "互联网/科技"], ["大疆", "互联网/科技"],
  ["海康威视", "互联网/科技"], ["商汤", "互联网/科技"], ["旷视", "互联网/科技"], ["科大讯飞", "互联网/科技"],
  ["apple", "互联网/科技"], ["microsoft", "互联网/科技"], ["google", "互联网/科技"], ["amazon", "互联网/科技"],
  ["meta", "互联网/科技"], ["nvidia", "互联网/科技"], ["amd", "互联网/科技"], ["intel", "互联网/科技"],
  // 汽车/出行
  ["比亚迪", "汽车/出行"], ["蔚来", "汽车/出行"], ["理想汽车", "汽车/出行"], ["小鹏", "汽车/出行"],
  ["吉利", "汽车/出行"], ["长城汽车", "汽车/出行"], ["广汽", "汽车/出行"], ["上汽", "汽车/出行"],
  ["一汽", "汽车/出行"], ["长安汽车", "汽车/出行"], ["奇瑞", "汽车/出行"], ["滴滴", "汽车/出行"],
  ["宁德时代", "能源/化工"], ["tesla", "汽车/出行"], ["特斯拉", "汽车/出行"],
  // 消费/零售
  ["农夫山泉", "消费/零售"], ["养生堂", "消费/零售"], ["蒙牛", "消费/零售"], ["伊利", "消费/零售"],
  ["海天", "消费/零售"], ["双汇", "消费/零售"], ["光明乳业", "消费/零售"], ["安踏", "消费/零售"],
  ["李宁", "消费/零售"], ["名创优品", "消费/零售"], ["欧莱雅", "消费/零售"], ["宝洁", "消费/零售"],
  ["联合利华", "消费/零售"], ["百事", "消费/零售"], ["可口可乐", "消费/零售"], ["雀巢", "消费/零售"],
  // 金融
  ["招商银行", "金融"], ["工商银行", "金融"], ["建设银行", "金融"], ["中国银行", "金融"],
  ["农业银行", "金融"], ["平安", "金融"], ["中国人寿", "金融"], ["太平洋保险", "金融"],
  ["微众银行", "金融"], ["陆金所", "金融"], ["华夏银行", "金融"],
  // 医疗/医药
  ["恒瑞", "医疗/医药"], ["药明康德", "医疗/医药"], ["迈瑞", "医疗/医药"], ["复星医药", "医疗/医药"],
  ["智飞", "医疗/医药"], ["百济神州", "医疗/医药"], ["pfizer", "医疗/医药"], ["novartis", "医疗/医药"],
  ["roche", "医疗/医药"], ["astrazeneca", "医疗/医药"],
  // 能源/化工
  ["隆基", "能源/化工"], ["中石油", "能源/化工"], ["中石化", "能源/化工"], ["国家电网", "能源/化工"],
  ["协鑫", "能源/化工"], ["万华化学", "能源/化工"],
  // 制造/工业
  ["富士康", "制造/工业"], ["立讯精密", "制造/工业"], ["三一", "制造/工业"], ["潍柴", "制造/工业"],
  ["中联重科", "制造/工业"], ["徐工", "制造/工业"], ["siemens", "制造/工业"], ["西门子", "制造/工业"],
  ["abb", "制造/工业"], ["施耐德", "制造/工业"], ["博世", "汽车/出行"],
  // 物流/供应链
  ["顺丰", "物流/供应链"], ["货拉拉", "物流/供应链"], ["中通", "物流/供应链"], ["圆通", "物流/供应链"],
  ["菜鸟", "物流/供应链"], ["京东物流", "物流/供应链"], ["满帮", "物流/供应链"],
  // 地产/建筑
  ["万科", "地产/建筑"], ["保利", "地产/建筑"], ["中建", "地产/建筑"], ["碧桂园", "地产/建筑"],
  ["龙湖", "地产/建筑"],
  // 传媒/文娱
  ["爱奇艺", "传媒/文娱"], ["芒果", "传媒/文娱"], ["阅文", "传媒/文娱"], ["米哈游", "互联网/科技"],
  // 教育
  ["新东方", "教育"], ["好未来", "教育"], ["学而思", "教育"],
];

// 行业关键词规则（公司名含该词 → 行业）。顺序敏感：更具体/易误判的在前。
// 命中即返回，故把「汽车/医药/金融」等强信号放在「制造/科技」等泛信号之前，避免被泛词抢先。
const INDUSTRY_KEYWORD_RULES = [
  ["金融", /银行|证券|保险|基金|信托|期货|资管|财险|寿险|金融|支付|消费金融|小额贷|bank|securities|insurance|capital/i],
  ["医疗/医药", /医药|制药|药业|药品|生物医药|生物科技|医疗|医院|健康|基因|诊断|器械|pharma|biotech|medical|health/i],
  ["汽车/出行", /汽车|整车|车业|新能源车|乘用车|商用车|车联网|出行|motors|automotive/i],
  ["能源/化工", /能源|电力|电网|石油|石化|化工|化学|新能源|光伏|风电|储能|电池|燃气|煤业|核电|energy|power|chemical|petro/i],
  ["物流/供应链", /物流|快递|供应链|仓储|货运|运输|冷链|logistics|express|supply\s*chain/i],
  ["地产/建筑", /地产|置业|房产|建筑|建设|建工|工程局|装饰|幕墙|real\s*estate|construction|properties/i],
  ["教育", /教育|学校|培训|学院|课程|留学|education|academy/i],
  ["传媒/文娱", /传媒|影视|文化|娱乐|院线|音乐|动漫|文娱|出版|media|entertainment/i],
  ["消费/零售", /食品|饮料|乳业|乳品|零售|商超|百货|便利店|美妆|化妆品|日化|服饰|服装|鞋业|家居|家电|餐饮|连锁|消费|快消|retail|consumer|foods?|beverage/i],
  ["制造/工业", /制造|机械|重工|工业|装备|设备|电子|半导体|芯片|集成电路|材料|钢铁|有色|精密|模具|纺织|轻工|manufactur|industrial|electronics|semiconductor/i],
  ["互联网/科技", /互联网|科技|网络|信息技术|软件|数码|智能|大数据|云计算|游戏|网游|人工智能|物联网|tech|software|digital|internet|\bai\b|cloud/i],
];

// 公司 → 行业类目（或 null=判不出）。
function classifyCompanyIndustry(company) {
  const text = normalizeCompany(company);
  if (!text) return null;
  for (const [name, cat] of COMPANY_OVERRIDES) {
    if (text.includes(normalizeCompany(name))) return cat;
  }
  for (const [cat, rule] of INDUSTRY_KEYWORD_RULES) {
    if (rule.test(text)) return cat;
  }
  return null;
}

// 用户自填行业（自由文本，如「互联网」「快消」「生物医药」）→ 归一到 INDUSTRY_CATEGORIES 之一（或 null）。
// 用别名/子串映射，让用户的口语化输入对齐分类器类目空间。
const USER_INDUSTRY_ALIASES = [
  ["互联网/科技", /互联网|科技|信息技术|软件|计算机|it|tech|游戏|人工智能|\bai\b|大数据|云/i],
  ["金融", /金融|银行|证券|保险|基金|投资|fintech|finance/i],
  ["消费/零售", /消费|零售|快消|fmcg|电商|食品|饮料|美妆|服装|retail|consumer/i],
  ["制造/工业", /制造|工业|机械|电子|半导体|芯片|材料|硬件|manufactur|industrial/i],
  ["汽车/出行", /汽车|车|出行|新能源车|automotive/i],
  ["医疗/医药", /医疗|医药|生物|制药|健康|器械|pharma|bio|medical|health/i],
  ["能源/化工", /能源|电力|化工|化学|新能源|光伏|电池|energy|chemical/i],
  ["地产/建筑", /地产|房地产|建筑|建设|工程|real\s*estate|construction/i],
  ["物流/供应链", /物流|供应链|快递|运输|logistics|supply/i],
  ["传媒/文娱", /传媒|文娱|影视|文化|娱乐|内容|media|entertainment/i],
  ["教育", /教育|培训|edu/i],
];

function canonicalizeUserIndustry(value) {
  const text = normalizeCompany(value);
  if (!text) return null;
  // 已是规范类目则直接用。
  if (INDUSTRY_CATEGORIES.includes(value)) return value;
  for (const [cat, rule] of USER_INDUSTRY_ALIASES) {
    if (rule.test(text)) return cat;
  }
  return null;
}

// 把用户自填行业数组归一成规范类目集合（去 null/去重）。
function userTargetIndustryCategories(userIndustries) {
  const out = new Set();
  for (const raw of userIndustries || []) {
    const cat = canonicalizeUserIndustry(raw);
    if (cat) out.add(cat);
  }
  return out;
}

// 跨行业门判定：用户岗位是否「行业相容」。
// 放行（返回 true）当：用户没填可识别行业 / 岗位行业判不出（缺数据不误杀）/ 行业落在用户目标集合内。
// 拦截（返回 false）仅当：用户有明确目标行业 且 岗位行业已知 且 不在目标集合内（治「同职能跨行业」误命中）。
function jobIndustryAllowed(company, userIndustries) {
  const targets = userTargetIndustryCategories(userIndustries);
  if (targets.size === 0) return true;
  const jobCat = classifyCompanyIndustry(company);
  if (!jobCat) return true;
  return targets.has(jobCat);
}

module.exports = {
  INDUSTRY_CATEGORIES,
  classifyCompanyIndustry,
  canonicalizeUserIndustry,
  userTargetIndustryCategories,
  jobIndustryAllowed,
};
